crates/driver/src/domain/competition/solution/scoring.rs

```diff
@@ -0,0 +1,262 @@
+                        // with the surplus
+                        let executed_in_surplus_token = match self.side {
+                            Side::Sell => self
+                                .executed
+                                .0
+                                .checked_mul(self.custom_price.sell)
+                                .ok_or(super::Math::Overflow)?
+                                .checked_div(self.custom_price.buy)
+                                .ok_or(super::Math::DivisionByZero)?,
+                            Side::Buy => self
+                                .executed
+                                .0
+                                .checked_mul(self.custom_price.buy)
+                                .ok_or(super::Math::Overflow)?
+                                .checked_div(self.custom_price.sell)
+                                .ok_or(super::Math::DivisionByZero)?,
+                        };
+                        let factor = match self.side {
+                            Side::Sell => max_volume_factor / (1.0 - max_volume_factor),
+                            Side::Buy => max_volume_factor / (1.0 + max_volume_factor),
+                        };
+                        apply_factor(executed_in_surplus_token, factor)
+                            .ok_or(Error::Factor(executed_in_surplus_token, factor))?
+                    },
+                )),
+                order::FeePolicy::PriceImprovement {
+                    factor: _,
+                    max_volume_factor: _,
+                    quote: _,
+                } => Err(Error::UnimplementedFeePolicy),
+                order::FeePolicy::Volume { factor: _ } => Err(Error::UnimplementedFeePolicy),
+            }
+        };
+
+        let protocol_fee = self.policies.first().map(protocol_fee).transpose();
+        Ok(eth::Asset {
+            token: self.surplus_token(),
+            amount: protocol_fee?.unwrap_or(0.into()).into(),
+        })
+    }
+
+    /// Protocol fee is defined by fee policies attached to the order.
+    ///
+    /// Denominated in NATIVE token
+    fn native_protocol_fee(&self, prices: &auction::Prices) -> Result<eth::TokenAmount, Error> {
+        let protocol_fee = self
+            .surplus_token_price(prices)?
+            .apply(self.protocol_fee()?.amount);
+        // normalize
+        Ok((protocol_fee.0 / *UNIT).into())
+    }
+
+    fn surplus_token(&self) -> eth::TokenAddress {
+        match self.side {
+            Side::Buy => self.sell.token,
+            Side::Sell => self.buy.token,
+        }
+    }
+
+    /// Returns the price of the trade surplus token
+    fn surplus_token_price(&self, prices: &auction::Prices) -> Result<auction::Price, Error> {
+        prices
+            .get(&self.surplus_token())
+            .cloned()
+            .ok_or(Error::MissingPrice(self.surplus_token()))
+    }
+}
+
+fn apply_factor(amount: eth::U256, factor: f64) -> Option<eth::U256> {
```

Comment:
Can this be implemented on `TokenAmount` or some other shared domain object to make sure we have the same logic for applying factors to token amounts across the domain (and document our considerations)?